### ISSUE TITLE
Non-deterministically corrupted results from np.dot (conditional on large imports)

### ISSUE THREAD

**Issue by ilya-kolpakov** [numpy/numpy#12394]
I am getting randomly corrupted results from `np.dot` if the first argument is an array of doubles with FORTRAN layout.

The behavior is non-deterministic and the does not occur unless I import a large module beforehand (e.g. `seaborn` or `theano` does the trick).

### Reproducing code example:

```python
import numpy as np

random = dict(
    int = lambda shape: np.random.randint(low=0, high=10, size=shape),
    float = lambda shape: np.random.randn(*shape)
)

def check_dot(A, b, attempts=10, eps=1e-1):
    expected = (A * b).sum(axis=1)
    for i in range(attempts):
        actual = np.dot(A, b)
        if not np.allclose(expected, actual, eps):
            return False
    return True

def test(n_cols=2):
    for n_rows in 2**np.arange(1, 16, 1):
        for type, f in random.items():
            A = f((n_rows, n_cols))
            b = f((n_cols,))
            for order in 'CF':
                Ac = np.copy(A, order=order)
                if not check_dot(Ac, b):
                    print(type, n_rows, order)
                    return Ac, b
    return None


assert test() is None
import seaborn # or import theano
Ac, b = test()
```
prints
```
float 8192 F
```
The error reliably occurs for matrices of larger sizes.

### Examples of corrupted results (random data)

If the failure occurs and I evaluate `np.dot(Ac, b)` a few times 
```
import matplotlib.pyplot as plt

n_eval = 4
f, axes = plt.subplots(n_eval, 1, True, True, figsize=(8, 1.5*n_eval))

expected = (Ac*b).sum(axis=1)
for ax in axes:
    actual  = np.dot(Ac, b)
    ax.plot(actual - expected)

```
the output is non-deterministic and the errors look like this:
![errors-randn](https://user-images.githubusercontent.com/592849/48546769-71326a00-e8c9-11e8-83f5-5e9bbf42e0f6.png)

### Examples of corrupted results (structured data)

If I replace the normal random generator with `linspace`:
```python
random = dict(
    int = lambda shape: np.random.randint(low=0, high=10, size=shape),
    float = lambda shape: np.linspace(0, 1, np.product(shape)).reshape(shape)
)
```
I get errors plots looking like this:
![errors-linspace](https://user-images.githubusercontent.com/592849/48548017-3251e380-e8cc-11e8-88b1-396335361056.png)

### Numpy/Python version information:
Since I am using numpy packaged for [nix](www.nixos.com) package manager it is easy to get exactly the same Python environment package on a different machine. I did not do this yet.

```
>> sys.version
3.6.6 (default, Jun 27 2018, 05:47:41) 
[GCC 7.3.0]

>> np.__version__
1.15.1

>> np.show_config()
blas_mkl_info:
  NOT AVAILABLE
blis_info:
  NOT AVAILABLE
openblas_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/nix/store/blk28p4cr6r2nc7fi1c4gggiqpd7pkqy-openblas-0.3.1/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
blas_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/nix/store/blk28p4cr6r2nc7fi1c4gggiqpd7pkqy-openblas-0.3.1/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
lapack_mkl_info:
  NOT AVAILABLE
openblas_lapack_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/nix/store/blk28p4cr6r2nc7fi1c4gggiqpd7pkqy-openblas-0.3.1/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]
lapack_opt_info:
    libraries = ['openblas', 'openblas']
    library_dirs = ['/nix/store/blk28p4cr6r2nc7fi1c4gggiqpd7pkqy-openblas-0.3.1/lib']
    language = c
    define_macros = [('HAVE_CBLAS', None)]

```

**Comment by charris** [comment]
Hmm, OpenBLAS 0.3.1 is known buggy, that is why we don't use it. Where did you get NumPy?

**Comment by ilya-kolpakov** [comment]
Thanks a lot!

I'm using [nix package manager](https://nixos.org). The latest stable release of packages (18.09) happened to have OpenBLAS [0.3.1](https://github.com/NixOS/nixpkgs/blob/9bf8b0d06b7a0992a20abcc4a9cd68d9ebb4bad4/pkgs/development/libraries/science/math/openblas/default.nix#L82) as a default version (upstream is already at [0.3.3](https://github.com/NixOS/nixpkgs/blob/101133deaeea1194f4a915d8f27ad485d68e4e36/pkgs/development/libraries/science/math/openblas/default.nix#L82ent/libraries/science/math/openblas/default.nix#L82)).

I created an [issue](https://github.com/NixOS/nixpkgs/issues/50407) there where I am arguing that by default nix should package numpy/scipy using exactly the same versions of dependencies as you do in your binary builds. Do I understand correctly that you are building numpy `1.15.x` with OpenBLAS 0.3.0 as reported in the [release notes](https://github.com/numpy/numpy/blob/master/doc/release/1.15.0-notes.rst)?

**Comment by charris** [comment]
Yes, we built the wheels with 0.3.0. See comments https://github.com/MacPython/numpy-wheels/commit/cd530708178788aa3f6edb2df4acbc1daf29af69 . Note that we have discovered that 0.3.0 is not thread safe, so will be looking for 0.3.4.

**Comment by charris** [comment]
See https://github.com/xianyi/OpenBLAS/issues/1851 for 0.3.0 threading problem.

**Comment by charris** [comment]
See also https://github.com/xianyi/OpenBLAS/issues/1844.

**Comment by ilya-kolpakov** [comment]
Thanks. Apparently it's indeed multithreading-related. I get no errors if I force OpenBLAS to use only one thread. I don't think seaborn creates any threads so it's not clear why import matters.

**Comment by brada4** [comment]
You could try to apply [diff](https://patch-diff.githubusercontent.com/raw/xianyi/OpenBLAS/pull/1865.diff) from PR referenced in linked issue that is supposed to fix it. It is quite old code being modified, thus should apply cleanly on many even not so recent versions. It is due for 0.3.4 if you can wait a bit and upgrade binary.

**Comment by charris** [comment]
@brada4 What sort of schedule are you looking at for 0.3.4. I'm looking to delay NumPy 1.16 until it comes out.

**Comment by brada4** [comment]
@martin-frbg has plan exactly because of this issue:
https://github.com/xianyi/OpenBLAS/pull/1865#issuecomment-437961004

**Comment by charris** [comment]
I put a 1.16 milestone on this for tracking purposes.

**Comment by charris** [comment]
I think this has been fixed in OpenBLAS 3.4, so removing the milestone. @ilya-kolpakov you can test this now by downloading the latest numpy wheel builds from https://7933911d6844c6c53a7d-47bd50c35cd79bd838daf386af554a83.ssl.cf2.rackcdn.com/, look for files beginning `numpy-1.16.0.dev0+20181204`, or you can wait for 1.16.0rc1.

**Comment by seberg** [comment]
Considering that this is an OpenBLAS issue from more then half a year ago, which apparently was fixed upstream. Closing this issue.
